### PR TITLE
Kernel/Ext2FS: Add full support for large inodes

### PR DESCRIPTION
**Pull request by implicitfield**
The extra data provided by large inodes has almost exclusively to do with timestamps, which means that you can now get nanosecond precision on your root file system, and the capability to support dates beyond 2038 or so in a standards-compliant way. This is also basically a pre-requisite for ext3/ext4, since the 128-byte inodes we've been using are more-or-less deprecated.

This PR also fixes up serialization of epoch offsets, so if you have a filesystem that has inodes with negative timestamps, then those should now be reported correctly.

Lastly, this also adds a CI step to validate filesystem consistency after running tests, since we shouldn't be seeing any filesystem corruption after all the recent work that's been put into `Ext2FS`.

### CODE REVIEW

File: Kernel/FileSystem/Ext2FS/Inode.cpp [hunk 1]
```diff
@@ -895,18 +933,30 @@ ErrorOr<void> Ext2FSInode::update_timestamps(Optional<UnixDateTime> atime, Optio
...
     if (fs().is_readonly())
         return EROFS;
-    if (atime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max())
+    if (atime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max() && !is_within_inode_bounds(reinterpret_cast<FlatPtr>(&m_raw_inode), reinterpret_cast<FlatPtr>(&m_raw_inode.i_atime_extra), sizeof(m_raw_inode.i_atime_extra)))
         return EINVAL;
-    if (ctime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max())
+    if (ctime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max() && !is_within_inode_bounds(reinterpret_cast<FlatPtr>(&m_raw_inode), reinterpret_cast<FlatPtr>(&m_raw_inode.i_ctime_extra), sizeof(m_raw_inode.i_ctime_extra)))
         return EINVAL;
-    if (mtime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max())
+    if (mtime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max() && !is_within_inode_bounds(reinterpret_cast<FlatPtr>(&m_raw_inode), reinterpret_cast<FlatPtr>(&m_raw_inode.i_mtime_extra), sizeof(m_raw_inode.i_mtime_extra)))
```
Comment (Hendiadyoin1) [inline Kernel/FileSystem/Ext2FS/Inode.cpp:940]: please prefer `bit_cast` over `reinterpret_cast` here

**Comment by nico** [comment]
I currently rely on genext2fs. Will that still work after this PR? One of the commit messages sound a bit ominous :)

File: Kernel/FileSystem/Ext2FS/Inode.cpp [hunk 2]
```diff
@@ -417,10 +427,38 @@ InodeMetadata Ext2FSInode::metadata() const
...
+    // fact that we always store the raw inode as an ext2_inode_large.
+    if (is_within_inode_bounds(bit_cast<FlatPtr>(&m_raw_inode), bit_cast<FlatPtr>(&m_raw_inode.i_atime_extra), sizeof(m_raw_inode.i_atime_extra))) {
+        time_t seconds = decode_seconds_with_extra(m_raw_inode.i_atime, m_raw_inode.i_atime_extra);
+        u32 nanoseconds = decode_nanoseconds_from_extra(m_raw_inode.i_atime_extra);
+        metadata.atime = UnixDateTime::from_unix_timespec({ .tv_sec = seconds, .tv_nsec = nanoseconds });
+    } else {
+        metadata.atime = UnixDateTime::from_seconds_since_epoch(static_cast<i32>(m_raw_inode.i_atime));
+    }
+
+    // NOTE: There's no i_dtime_extra, so we use i_ctime_extra to approximate the right epoch for metadata.dtime (like e2fsprogs.)
```
Comment (nico) [inline Kernel/FileSystem/Ext2FS/Inode.cpp:442]: e2fsprogs is GPL I think, so we shouldn't have comments that sound like we looked at its source. Maybe this is easy to see from its external behavior?

File: Kernel/FileSystem/Ext2FS/Inode.cpp [hunk 3]
```diff
@@ -895,18 +933,30 @@ ErrorOr<void> Ext2FSInode::update_timestamps(Optional<UnixDateTime> atime, Optio
...
-        m_raw_inode.i_ctime = ctime.value().to_timespec().tv_sec;
-    if (mtime.has_value())
-        m_raw_inode.i_mtime = mtime.value().to_timespec().tv_sec;
+    if (atime.has_value()) {
+        time_t seconds = atime.value().to_timespec().tv_sec;
+        u32 nanoseconds = atime.value().to_timespec().tv_nsec;
+        m_raw_inode.i_atime = static_cast<u32>(seconds);
+        m_raw_inode.i_atime_extra = encode_time_to_extra(seconds, nanoseconds);
+    }
+    if (ctime.has_value()) {
```
Comment (nico) [inline Kernel/FileSystem/Ext2FS/Inode.cpp:948]: This feels pretty repetitive. Could be shorter with a local lambda for the common code. (Further up too, to a lesser degree.)

**Review by nico — APPROVED** [review]
Looks vaguely right-shaped :)

**Pull request merged**